gh-124213: Skip tests failing with ``--suppress-sync=true`` when inside systemd-nspawn container

mgorny opened the pull request:
Add a helper functon that checks whether the test suite is running inside a systemd-nspawn container, and skip the few tests failing with `--suppress-sync=true` in that case.  The tests are failing because `--suppress-sync=true` stubs out `fsync()`, `fdatasync()` and `msync()` calls, and therefore they always return success without checking for invalid arguments.

Technically, this means that the tests are also skipped when running inside systemd-nspawn container without `--suppress-sync=true`. However, to the best of my knowledge the only way to detect this option is to actually reproduce one of the unexpectedly-succeeding syscalls, and since this is precisely what we're testing for, the skip-test and the actual test would be doing the same thing.

<!-- gh-issue-number: gh-124213 -->
* Issue: gh-124213
<!-- /gh-issue-number -->

mgorny committed:
gh-124213: Skip tests failing with ``--suppress-sync=true`` when inside systemd-nspawn container

Add a helper functon that checks whether the test suite is running
inside a systemd-nspawn container, and skip the few tests failing
with `--suppress-sync=true` in that case.  The tests are failing because
`--suppress-sync=true` stubs out `fsync()`, `fdatasync()` and `msync()`
calls, and therefore they always return success without checking for
invalid arguments.

Technically, this means that the tests are also skipped when running
inside systemd-nspawn container without `--suppress-sync=true`.
However, to the best of my knowledge the only way to detect this option
is to actually reproduce one of the unexpectedly-succeeding syscalls,
and since this is precisely what we're testing for, the skip-test
and the actual test would be doing the same thing.

vstinner commented on Lib/test/support/__init__.py:
```diff
@@ -2873,3 +2875,11 @@ def __iter__(self):
         return self
+
+
+def in_systemd_nspawn() -> bool:
```
Please move `from pathlib import Path` here to avoid importing pathlib in all tests using test.support (reduce the number of modules imported by tests).

Moreover, I'm not sure that pathlib dependency is useful here. You can just with:

```
with open("/run/systemd/container", "rb") as fp:
    container = fp.read()
return (container.rstrip() == b"systemd-nspawn")
```

vstinner replied:
Please explain a little bit what is nspawn and what is the purpose of this function in a docstring or a comment.

vstinner commented on Misc/NEWS.d/next/Tests/2024-09-18-18-39-21.gh-issue-124213.AQq_xg.rst:
```diff
@@ -0,0 +1,2 @@
+Skip tests failing with ``--suppress-sync=true`` when running inside
```
```suggestion
Skip test_os and test_mmap tests failing with ``--suppress-sync=true`` when running inside
```

mgorny committed:
Implement review comments from @vstinner

Add a docstring explaining the function in detail.  Switch from pathlib
to plain `open().read()`.  Update the news entry.

mgorny committed:
Detect `--suppress-sync=true` and limit the skips to that

Call `os.open("", os.O_RDONLY | os.O_SYNC)` and check the errno to
detect whether `--suppress-sync=true` is actually used, and skip
the tests only in that scenario.

mgorny committed:
Fix formatting (typo) in the news entry

vstinner commented on Lib/test/support/__init__.py:
```diff
@@ -2873,3 +2874,40 @@ def __iter__(self):
+        # If systemd-nspawn is used, O_SYNC flag will immediately trigger
+        # EINVAL.  Otherwise, ENOENT will be given instead.
+        try:
+            with os.open("", os.O_RDONLY | os.O_SYNC):
```
It's strange to pass an empty path. I suggest to use ``__file__`` instead:

```suggestion
            with os.open(__file__, os.O_RDONLY | os.O_SYNC):
```

vstinner commented on Lib/test/support/__init__.py:
```diff
@@ -2873,3 +2874,40 @@ def __iter__(self):
+    Return True if the test suite is being run inside a systemd-nspawn
+    container and ``--suppress-sync=true`` option is detected to be used,
+    False otherwise.  This can be used to skip tests that rely
+    on ``fsync()`` calls and similar not being intercepted.
```
The doc repeats itself, you can just say:

```suggestion
    This can be used to skip tests that rely on ``fsync()`` calls
    and similar not being intercepted.
```

vstinner commented on Lib/test/support/__init__.py:
```diff
@@ -2873,3 +2874,40 @@ def __iter__(self):
+
+def in_systemd_nspawn_sync_suppressed() -> bool:
+    """
+    Test whether the test suite is runing in systemd-nspawn with ``--suppress-sync=true``
```
```suggestion
    Test whether the test suite is runing in systemd-nspawn
    with ``--suppress-sync=true``.
```

vstinner commented on Lib/test/support/__init__.py:
```diff
@@ -2873,3 +2874,40 @@ def __iter__(self):
+        with open("/run/systemd/container", "rb") as fp:
+            return fp.read().rstrip() == b"systemd-nspawn"
+    except FileNotFoundError:
+        return False
```
Can you just inline this code in in_systemd_nspawn_sync_suppressed()? This function doesn't seem to be used outside in_systemd_nspawn_sync_suppressed().

mgorny replied:
Sure can do. My idea was that having it separate makes it self-explanatory, and potentially useful if we ever have any other problems with `systemd-nspawn`, but I'm fine either way.

mgorny committed:
Apply suggestions from @vstinner

Co-authored-by: Victor Stinner <vstinner@python.org>

mgorny committed:
Inline in_systemd_nspawn() in *_sync_suppressed()

vstinner commented on Lib/test/support/__init__.py:
```diff
@@ -2873,3 +2874,37 @@ def __iter__(self):
         return self
+
+
+
```
```suggestion


```

vstinner commented on Lib/test/support/__init__.py:
```diff
@@ -2873,3 +2874,37 @@ def __iter__(self):
+    import errno
+
+    # If systemd-nspawn is used, O_SYNC flag will immediately
+    # trigger EINVAL.  Otherwise, ENOENT will be given instead.
```
```suggestion
    # If systemd-nspawn is used, O_SYNC flag will immediately
    # trigger EINVAL.  Otherwise, ENOENT will be given instead.
    import errno
```

vstinner commented on Lib/test/test_mmap.py:
```diff
@@ -839,7 +840,7 @@ def test_flush_return_value(self):
         result = mm.flush()
         self.assertIsNone(result)
-        if sys.platform.startswith(('linux', 'android')):
+        if sys.platform.startswith(('linux', 'android')) and not in_systemd_nspawn_sync_suppressed():
```
```suggestion
        if (sys.platform.startswith(('linux', 'android')) 
            and not in_systemd_nspawn_sync_suppressed()):
```

vstinner committed:
Apply suggestions from code review

vstinner reviewed (approved):
LGTM

vstinner commented on Lib/test/test_mmap.py:
```diff
@@ -839,7 +840,8 @@ def test_flush_return_value(self):
         result = mm.flush()
         self.assertIsNone(result)
-        if sys.platform.startswith(('linux', 'android')):
+        if (sys.platform.startswith(('linux', 'android')) 
```
```suggestion
        if (sys.platform.startswith(('linux', 'android'))
```

vstinner committed:
Update Lib/test/test_mmap.py

mgorny commented on Lib/test/test_mmap.py:
```diff
@@ -839,7 +840,8 @@ def test_flush_return_value(self):
         self.assertIsNone(result)
-        if sys.platform.startswith(('linux', 'android')):
+        if (sys.platform.startswith(('linux', 'android'))
+            and not in_systemd_nspawn_sync_suppressed()):
```
To be honest, I wanted to avoid this because it's very confusing when the condition has the same indent as the code below.

vstinner replied:
Yeah but the line was too long. That's the problem of long function names.

mgorny commented on Lib/test/test_mmap.py:
```diff
@@ -839,7 +840,8 @@ def test_flush_return_value(self):
         result = mm.flush()
         self.assertIsNone(result)
-        if sys.platform.startswith(('linux', 'android')):
+        if (sys.platform.startswith(('linux', 'android'))
+            and not in_systemd_nspawn_sync_suppressed()):
```
Perhaps we could follow black here:

```suggestion
        if (
            sys.platform.startswith(("linux", "android"))
            and not in_systemd_nspawn_sync_suppressed()
        ):
```
